### ISSUE TITLE
Fn::Sub warning on Resources line

### ISSUE THREAD

**Issue by peterneave** [aws-cloudformation/cfn-lint#3440]
### CloudFormation Lint Version

1.4.2

### What operating system are you using?

Windows, Ububunu

### Describe the bug

I'm getting this warning from cfn-lint (1.30 to 1.4.2). - 0.87 is fine by the way. The problem is that it's for the `Resources:` line section. 

```sh
W1020 'Fn::Sub' isn't needed because there are no variables
src\project\project.yml:115:1
```

There are only two places where `Fn::Sub` is getting used in the whole file and I've checked them. I also checked `!Sub` references and from what I can see, they look fine too. Obviously having the actual line number that I should be looking at would make a big difference rather than the parent.

I won't include a template as it would be 4k lines. I'm sorry I can't be more specific about this issue as it's kind of vague in of itself.

### Expected behavior

I would expect to be directed to the exact line that it's warning me about - not the parent.

### Reproduction template

```yml
Resources:
  ...
```

**Comment by kddejong** [comment]
Any chance you have the resource/property you are using the sub?

I'll look into why the path is getting messed up.

That rule will look for the ${}(with a negative look ahead for the escape characters) pattern in the Fn::Sub string section to validate if there are any parameters. Can you double check that there are variables there.  Its possible that the regex is getting tripped on something but I don't believe this regex was changed from v0.

Also are you using SAM in this template?

**Comment by kddejong** [comment]
This is the regex by the way.  `re.compile(r"\${([^!].*?)}")`

**Comment by peterneave** [comment]
It is a SAM template. There are two AWS::CloudWatch::Dashboard's that use `Fn::Sub` and even if I remove those logical resources from the template, it still brings up the warning (i.e there is no `Fn::Sub` found in the file at all). That is what is odd about it.

**Comment by kddejong** [comment]
By the time we do validation all occurrences of `!Sub` are converted into `Fn::Sub`

**Comment by kddejong** [comment]
I'm curious if you were able to figure this out.  You should be able to do `cfn-lint --info -- template.yaml` to see the template that is generated by SAM translator.  We use that template to do our linting.  My guess is there a sub that doesn't have a parameter in there.  We have a few options... we could disable this validation when a person is using SAM.  The reason resources is being highlighted is its a new resource that is created by SAM so we lose the ability to track it to the template you are working on.

**Comment by peterneave** [comment]
Sorry for the delay - I was ill. I eventually go the output to a file with WSL and  `cfn-lint --info --  ./src/ropp/ropp.yml 2> output`. Unfortunately, the line at 115 it is complaining about is `  },` and the `Fn::Sub`'s at 109 and 120 have `${AWS::StackName}...`

We might just have to disable this warning to get past the issue at this stage.

**Comment by kddejong** [comment]
This error should no longer show in SAM transformed templates.
